core/src/main/java/tech/pegasys/eth2signer/core/multikey/metadata/AbstractArtifactSignerFactory.java

```diff
@@ -0,0 +1,118 @@
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.eth2signer.core.multikey.metadata;
+
+import static java.lang.String.format;
+import static java.nio.charset.StandardCharsets.UTF_8;
+
+import tech.pegasys.signers.azure.AzureKeyVault;
+import tech.pegasys.signers.hashicorp.HashicorpConnection;
+import tech.pegasys.signers.hashicorp.HashicorpConnectionFactory;
+import tech.pegasys.signers.hashicorp.TrustStoreType;
+import tech.pegasys.signers.hashicorp.config.ConnectionParameters;
+import tech.pegasys.signers.hashicorp.config.KeyDefinition;
+import tech.pegasys.signers.hashicorp.config.TlsOptions;
+
+import java.io.FileNotFoundException;
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.Optional;
+
+import com.google.common.io.Files;
+import org.apache.tuweni.bytes.Bytes;
+
+public abstract class AbstractArtifactSignerFactory implements ArtifactSignerFactory {
+
+  final HashicorpConnectionFactory connectionFactory;
+  final Path configsDirectory;
+
+  public AbstractArtifactSignerFactory(
+      final HashicorpConnectionFactory connectionFactory, final Path configsDirectory) {
+    this.connectionFactory = connectionFactory;
+    this.configsDirectory = configsDirectory;
+  }
+
+  protected Bytes extractBytesFromVault(final AzureSigningMetadata metadata) {
+    final AzureKeyVault azureVault =
+        new AzureKeyVault(
+            metadata.getClientId(),
+            metadata.getClientSecret(),
+            metadata.getTenantId(),
+            metadata.getVaultName());
+
+    final Optional<String> secret = azureVault.fetchSecret(metadata.getSecretName());
+    if (secret.isEmpty()) {
+      throw new SigningMetadataException("secret '" + metadata.getSecretName() + "' doesn't exist");
+    }
+    return Bytes.fromHexString(secret.get());
+  }
+
+  protected Bytes extractBytesFromVault(final HashicorpSigningMetadata metadata) {
+    TlsOptions tlsOptions = null;
+    if (metadata.getTlsEnabled()) {
+      final Path knownServerFile = metadata.getTlsKnownServerFile();
+      if (knownServerFile == null) {
+        tlsOptions = new TlsOptions(Optional.empty(), null, null); // use CA Auth
+      } else {
+        final Path configRelativeKnownServerPath = makeRelativePathAbsolute(knownServerFile);
+        if (!configRelativeKnownServerPath.toFile().exists()) {
+          throw new SigningMetadataException(
+              String.format(
+                  "Known servers file (%s) does not exist.", configRelativeKnownServerPath));
+        }
+        tlsOptions = new TlsOptions(Optional.of(TrustStoreType.WHITELIST), knownServerFile, null);
+      }
+    }
```

Comment:
nit: you may extract this out in a seperate method ... allowing you to use `final TlsOption tlsOptions = buildTlsOptions(metaData)` or similar.

Reply:
or better `final Optional<TlsOptions> tlsOptions = buildTlsOptions(metadata);`

Reply:
done